data/src/main/java/com/microsoft/azure/kusto/data/BaseClient.java

```diff
@@ -0,0 +1,188 @@
+package com.microsoft.azure.kusto.data;
+
+import com.azure.core.http.*;
+import com.azure.core.util.Context;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.microsoft.azure.kusto.data.exceptions.*;
+import com.microsoft.azure.kusto.data.http.CloseParentResourcesStream;
+import com.microsoft.azure.kusto.data.http.HttpRequestBuilder;
+import com.microsoft.azure.kusto.data.http.HttpStatus;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.http.conn.EofSensorInputStream;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.lang.invoke.MethodHandles;
+import java.util.Optional;
+import java.util.zip.DeflaterInputStream;
+import java.util.zip.GZIPInputStream;
+
+public abstract class BaseClient implements Client, StreamingClient {
+
+    private static final int MAX_REDIRECT_COUNT = 1;
+
+    // Make logger available to implementations
+    protected static final Logger LOGGER = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+
+    private final HttpClient httpClient;
+
+    public BaseClient(HttpClient httpClient) {
+        this.httpClient = httpClient;
+    }
+
+    protected String post(HttpRequest request) throws DataServiceException {
+
+        // Todo: Add async version of this method
+
+        // Execute and get the response
+        try (HttpResponse response = httpClient.sendSync(request, Context.NONE)) {
+            Optional<HttpHeader> contentEncoding = Optional.ofNullable(response.getHeaders().get(HttpHeaderName.CONTENT_ENCODING));
+            String responseBody = contentEncoding.isPresent() && (contentEncoding.get().getValue().contains("gzip")) ? response.getBodyAsInputStream()
```

Comment:
Thought about it, I see we're handling gzip/deflate headers 3 times in 3 different ways (here, next method, cloudinfo). Let's extract this handling to a common code.

Reply:
It's not perfect but these 3 should now be a little more aligned and I created and documented two static methods in Utils for identifying gzip compressed responses and getting specific stream implementations from a response stream.
Commit is[ here.](https://github.com/Azure/azure-kusto-java/pull/335/commits/5b6e2f7802384d4fd1363a443b087dbbdbff79d6)